Add metadata change support for processors

rdner opened the pull request:
## What does this PR do?

Some of the processors that are configured with a target field didn't
support special cases such as when the target field is `@timestamp` or a field in `@metadata`. This change is to make it consistent across all processors or to document why a particular processor does not need this change.

This PR continues the work done in https://github.com/elastic/beats/pull/30092

<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## Why is it important?

Since some processors (e.g. `fingerprint`) support setting `@metadata.*` fields the users can expect the same behaviour from the rest of the processors too. To avoid this confusion we must either support the same behaviour across all the processors or document clearly when it's not supported.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
~~- [ ] I have made corresponding change to the default configuration files~~
~~- [ ] I have added tests that prove my fix is effective or that my feature works~~ (existing tests pass, the existing behaviour should not change)
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Current state of all processors (after this change)

### Are not supposed to access `Meta` (have hard-coded field names):

* `./add_docker_metadata/add_docker_metadata.go` (sets `container.*` fields only)
* `./add_locale/add_locale.go` (the doc states "The processor adds the a `event.timezone` value to each event.")
* `./add_kubernetes_metadata/kubernetes.go` (sets `kubernetes.*` fields only)
* `./add_host_metadata/add_host_metadata.go` (sets `host.*` fields only)
* `./add_observer_metadata/add_observer_metadata.go` (sets `obverver.*` fields only)
* `./actions/include_fields.go` (the doc states "The `@timestamp` and type fields are always exported, even if they are not defined in the include_fields list.", so should be expected behaviour)
* `./actions/add_tags.go` (@kvch and I agreed that tags should be set only in fields and not metadata)
* `./actions/add_labels.go` (sets the `labels` field)
* `./add_cloud_metadata/add_cloud_metadata.go` (sets only cloud.* fields, not configurable to set `@metadata`)

### Do not modify anything:

* `./actions/drop_event.go`
* `./ratelimit/rate_limit.go`

### Use event-specific functions (support metadata change):
* `./actions/drop_fields.go` 
* `./actions/decode_base64_field.go`
* `./actions/decode_json_fields.go`
* `./actions/extract_field.go`
* `./actions/add_network_direction.go`
* `./actions/decompress_gzip_field.go`
* `./timestamp/timestamp.go`
* `./dns/dns.go`
* `./fingerprint/fingerprint.go`
* `./translate_sid/translatesid.go`
* `./add_id/add_id.go`
* `./urldecode/urldecode.go`
* `./registered_domain/registered_domain.go`
* `./decode_xml/decode_xml.go`
* `./convert/convert.go`
* `./communityid/communityid.go`
* `./extract_array/extract_array.go`
* `./decode_csv_fields/decode_csv_fields.go`
* `./dissect/processor.go`
* `./decode_xml_wineventlog/processor.go`
* `./script/processor.go`
* **`./actions/add_fields.go` [fixed in #30092]**
* **`./add_process_metadata/add_process_metadata.go` [fixed in this PR]**
* **`./actions/detect_mime_type.go` [fixed in this PR]**
* **`./actions/truncate_fields.go` [fixed in this PR]**
* **`./actions/replace.go` [fixed in this PR]**
* **`./actions/rename.go` [fixed in this PR]**
* **`./actions/copy_fields.go` [fixed in this PR]**

## Related issues

- Relates #30092
- Closes #25425

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane  (Team:Elastic-Agent-Data-Plane)

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30183/detail/PR-30183/2//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30183/detail/PR-30183/2//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30183/detail/PR-30183/2//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-30183/detail/PR-30183/2//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_30183.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-02-03T12:56:16.684Z&rangeTo=2022-02-03T13:16:16.684Z&transactionName=Beats/beats/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=c4ddde41e4e821db5f5142eaba02c80b&transactionId=844894c01d309041)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats



* Reason: `null`


* Start Time: 2022-02-03T13:06:16.684+0000


* Duration: 113 min 41 sec


* Commit: c234fe5a05b4019405b4d7f5d8bad4a294e5204a



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 48139  |
| Skipped      | 4292 |
| Total        | 52431   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












## :green_heart: Flaky test report
Tests succeeded.



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

cmacknz commented:
The changes themselves look good to me. The one concern I have is that we have no tests ensuring this inconsistency with the metadata fields won't happen again. Could we add a test for each affected processor to ensure that using `@metadata` actually modifies the metadata of the event and doesn't just add a new field?

The original reason this happened is that it isn't obvious at all from the code that the treatment of `@metadata` is different between the different types/methods for handling events. I would prefer if we can do something to enforce the correct behaviour in the future.

cmacknz reviewed (changes requested):
> The changes themselves look good to me. The one concern I have is that we have no tests ensuring this inconsistency with the metadata fields won't happen again. Could we add a test for each affected processor to ensure that using @metadata actually modifies the metadata of the event and doesn't just add a new field?
> 
> The original reason this happened is that it isn't obvious at all from the code that the treatment of @metadata is different between the different types/methods for handling events. I would prefer if we can do something to enforce the correct behaviour in the future.

Should have requested changes with this comment.

kvch commented on libbeat/processors/add_docker_metadata/add_docker_metadata.go:
```diff
@@ -132,6 +132,8 @@ func lazyCgroupCacheInit(d *addDockerMetadata) {
 }
 
+// Run runs the processor that adds container-related fields to the event
+// This processor does not modify any `Meta` fields in the event, it accesses and modifies `event.Fields` directly
```
Why did you change it? Do we win anything by touching processors that are not impacted?

rdner replied:
yes, now it looks more obvious that the processor is not allowed to edit metadata and works with the `Fields` only. I still think it's a good idea.

kvch commented on libbeat/processors/add_locale/add_locale.go:
```diff
@@ -89,7 +89,7 @@ func New(c *common.Config) (processors.Processor, error) {
 	zone, offset := time.Now().Zone()
 	format := l.Format(zone, offset)
-	event.PutValue("event.timezone", format)
+	event.Fields.Put("event.timezone", format)
```
TBH I prefer that in general no one updates the` Fields` attribute directly. I especially would not put extra effort into changing processors that are not impacted by the bug.

rdner replied:
Well, if there is a strong opinion against it – I can rollback the change. But I still think it's way easier to navigate the code this way. For example, when you find usages of `PutValue` you find all the occurrences where it's meant to be used with metadata support, all the occurrences of `Fields.Put` indicate no metadata support.

After reading all the 39 processors and looking for this distinction I find this useful :)

kvch replied:
I prefer to have only one way of extending events with new fields. `PutValue` is safe for both fields and metadata. Please roll back these changes.

> After reading all the 39 processors and looking for this distinction I find this useful :)

This way other developers do not have to read all of the 39 processors to find this pattern. :) We should simplify processor development so there are fewer possibilities for error. Maybe it is also worth adding a few sentences to our developer guide (if we have any about processor development).

rdner commented:
@cmacknz since it's quite a bit addition, I think it's a good idea to continue working on tests in a separate PR, what do you think?

cmacknz commented:
> @cmacknz since it's quite a bit addition, I think it's a good idea to continue working on tests in a separate PR, what do you think?

I'm not opposed to it, but I generally prefer when PRs come with tests proving they fixed what they say they did :) The existing tests tell us none of the existing functionality is broken, so in theory merging this won't break anything that isn't already broken (metadata field handling).

rdner commented:
@cmacknz I totally agree, it's just to do it right I need to introduce 24 new tests for each processor that is **supposed** to support changing metadata. I sure can continue in this PR, most likely will finish the tests tomorrow.

kvch reviewed (changes requested):
Please do not change existing `event.PutValue` to `event.Fields.Put`.

rdner commented:
@cmacknz I've added tests for the processors that support metadata changes:

* `./actions/drop_fields.go` 
* `./actions/decode_base64_field.go`
* `./actions/decode_json_fields.go`
* `./actions/extract_field.go`
* `./actions/add_network_direction.go`
* `./actions/decompress_gzip_field.go`
* `./timestamp/timestamp.go`
* `./dns/dns.go`
* `./fingerprint/fingerprint.go`
* `./translate_sid/translatesid.go`
* `./add_id/add_id.go`
* `./urldecode/urldecode.go`
* `./registered_domain/registered_domain.go`
* `./decode_xml/decode_xml.go`
* `./convert/convert.go`
* `./communityid/communityid.go`
* `./extract_array/extract_array.go`
* `./decode_csv_fields/decode_csv_fields.go`
* `./dissect/processor.go`
* `./decode_xml_wineventlog/processor.go`
* ~~`./script/processor.go`~~ (tests were already there https://github.com/elastic/beats/blob/8c79b67906ab4399dd9e575e5f2c4005498c32f3/libbeat/processors/script/javascript/beatevent_v0_test.go#L115-L127)
* ~~`./actions/add_fields.go`~~ (tests were added in #30092)
* `./add_process_metadata/add_process_metadata.go`
* `./actions/detect_mime_type.go`
* `./actions/truncate_fields.go`
* `./actions/replace.go`
* `./actions/rename.go`
* `./actions/copy_fields.go`


@kvch I removed the `event.PutValue` to `event.Fields.Put` changes.

cmacknz reviewed (approved):
Thanks, that's a lot of tests :)
